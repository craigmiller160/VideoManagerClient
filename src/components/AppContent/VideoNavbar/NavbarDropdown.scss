
.NavbarDropdown {
  .dropdown {
    background-color: #343a40;
    border: none;

    .dropdownItem {
      padding: .25rem 1.5rem;
    }
  }
}