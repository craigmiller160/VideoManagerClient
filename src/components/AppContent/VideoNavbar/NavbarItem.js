import React from 'react';
import PropTypes from 'prop-types';
import { NavItem, NavLink as BootLink } from 'reactstrap';
import { NavLink } from 'react-router-dom';
import classes from './NavbarItem.scss';
import useReactRouter from 'use-react-router';

const NavbarItem = (props) => {
    const { history: { pathname } } = useReactRouter();
    const {
        id,
        to,
        text,
        exact,
        isLink,
        onClick
    } = props;

    return (
        <NavItem className={ classes.NavbarItem } active={ to && pathname === to }>
            <BootLink
                tag="div"
                id={ id }
                className={ classes['use-pointer'] }
                onClick={ onClick }
            >
                {
                    isLink &&
                    <NavLink
                        to={ to }
                        activeClassName={ classes.active }
                        className={ classes.link }
                        exact={ exact }
                    >
                        { text }
                    </NavLink>
                }
                {
                    !isLink && text
                }
            </BootLink>
        </NavItem>
    );
};
NavbarItem.propTypes = {
    id: PropTypes.string,
    to: PropTypes.string,
    text: PropTypes.string,
    exact: PropTypes.bool,
    isLink: PropTypes.bool,
    onClick: PropTypes.func
};
NavbarItem.defaultProps = {
    exact: false,
    isLink: false
};

export default NavbarItem;
