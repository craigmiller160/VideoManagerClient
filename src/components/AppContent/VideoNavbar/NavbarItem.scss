@import './navbar';

.NavbarItem {
  .link {
    @include link;
  }
}
