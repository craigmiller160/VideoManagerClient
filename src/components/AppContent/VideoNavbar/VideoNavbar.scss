.VideoNavbar {
    border-bottom: 1px solid lightgray;
    .use-pointer {
        cursor: pointer;
    }
}