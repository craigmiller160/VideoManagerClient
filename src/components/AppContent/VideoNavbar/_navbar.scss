@mixin link {
  color: rgba(255,255,255,.5);
  &:hover {
    color: white;
    text-decoration: none;
  }
  &.active {
    color: white;
  }
}
