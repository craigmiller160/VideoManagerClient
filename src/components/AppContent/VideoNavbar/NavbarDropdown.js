/*
 *     VideoManagerClient
 *     Copyright (C) 2020 Craig Miller
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { DropdownMenu, DropdownToggle, UncontrolledDropdown } from 'reactstrap';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import useReactRouter from 'use-react-router';
import classes from './NavbarDropdown.scss';
import NavbarItem from './NavbarItem';
import { logout } from '../../../store/auth/auth.actions';

// TODO delete this, it's unnecessary

const NavbarDropdown = () => {
    const dispatch = useDispatch();
    const { location } = useReactRouter();
    const { firstName, lastName } = useSelector(state => state.auth.userDetails, shallowEqual);

    const toggleClasses = [];
    if (location.pathname === '/profile') {
        toggleClasses.push('active');
    }

    const fullName = `${firstName} ${lastName}`;

    return (
        <UncontrolledDropdown nav inNavbar className={ classes.NavbarDropdown }>
            <DropdownToggle id="vm-navbar-dropdown-toggle" nav caret className={ toggleClasses.join(' ') }>{ fullName }</DropdownToggle>
            <DropdownMenu id="vm-navbar-dropdown" right className={ classes.dropdown }>
                <NavbarItem
                    id="userProfileLink"
                    text="Profile"
                    className={ classes.dropdownItem }
                    isLink
                    exact
                    to="/profile"
                />
                <NavbarItem
                    id="logoutLink"
                    onClick={ () => dispatch(logout()) }
                    className={ classes.dropdownItem }
                    text="Logout"
                />
            </DropdownMenu>
        </UncontrolledDropdown>
    );
};

export default NavbarDropdown;
