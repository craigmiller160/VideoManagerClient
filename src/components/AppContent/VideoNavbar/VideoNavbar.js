import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Collapse, Container, Nav, Navbar, NavbarBrand, NavbarToggler } from 'reactstrap';
import * as classes from './VideoNavbar.scss';
import { NavLink } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import useReactRouter from 'use-react-router';
import { startFileScan } from '../../../store/scanning/scanning.actions';
import NavbarItem from './NavbarItem';
import NavbarDropdown from './NavbarDropdown';

const VideoNavbar = (props) => {
    const dispatch = useDispatch();
    const { history } = useReactRouter();
    const [ isOpen, setOpen ] = useState(false);
    const { disabled } = props;

    const onScanDirClick = async () => {
        await dispatch(startFileScan());
        history.push('/scanning');
    };

    return (
        <Navbar className={ classes.VideoNavbar } color="dark" dark expand="md">
            <Container>
                <NavbarBrand
                    tag="div"
                    disabled={ disabled }
                >
                    <NavLink
                        to="/"
                        activeClassName={ classes.active }
                        className={ classes.link }
                    >
                        Video Manager
                    </NavLink>
                </NavbarBrand>
                {
                    !disabled &&
                    <>
                        <NavbarToggler onClick={ () => setOpen(!isOpen) } />
                        <Collapse isOpen={ isOpen } navbar>
                            <Nav navbar>
                                <NavbarItem
                                    id="videoListLink"
                                    to="/list"
                                    exact
                                    text="Video List"
                                    isLink
                                />
                                <NavbarItem
                                    id="manageFiltersLink"
                                    to="/filters"
                                    exact
                                    text="Manage Filters"
                                    isLink
                                />
                            </Nav>
                            <Nav className="ml-auto" navbar>
                                <NavbarItem
                                    id="scanDirectoryLink"
                                    onClick={ onScanDirClick }
                                    text="Scan Directory"
                                />
                                <NavbarDropdown />
                            </Nav>
                        </Collapse>
                    </>
                }
            </Container>
        </Navbar>
    );
};

VideoNavbar.defaultProps = {
    disabled: false
};
VideoNavbar.propTypes = {
    disabled: PropTypes.bool
};

export default VideoNavbar;
