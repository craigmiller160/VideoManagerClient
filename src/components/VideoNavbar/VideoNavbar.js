import React, { Component } from 'react';
import {
    Nav,
    Navbar,
    NavbarBrand,
    NavItem,
    NavbarToggler,
    Collapse,
    NavLink
} from 'reactstrap';
import * as classes from './VideoNavbar.scss';

class VideoNavbar extends Component {

    state = {
        isOpen: false
    };

    toggle = () => {
        this.setState((prevState) => {
            return {
                ...prevState,
                isOpen: !prevState.isOpen
            }
        });
    };

    render() {

        return (
            <Navbar className={classes.VideoNavbar} color="light" light expand="md">
                <NavbarBrand href="/">Video Manager</NavbarBrand>
                <NavbarToggler onClick={this.toggle} />
                <Collapse isOpen={this.state.isOpen} navbar>
                    <Nav className="ml-auto" navbar>
                        <NavItem>
                            <NavLink>Scan Directory</NavLink>
                        </NavItem>
                    </Nav>
                </Collapse>
            </Navbar>
        );
    }

}

export default VideoNavbar;