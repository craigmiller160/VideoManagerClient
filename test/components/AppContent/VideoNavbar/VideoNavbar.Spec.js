import React from 'react';
import { mount } from 'enzyme';
import VideoNavbar from 'components/AppContent/VideoNavbar/VideoNavbar';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import useReactRouter from 'use-react-router'; // eslint-disable-line import/first

jest.mock('store/scanning/scanning.actions', () => ({
    startFileScan: () => ({ type: 'startFileScan' })
}));
jest.mock('store/auth/auth.actions', () => ({
    logout: () => ({ type: 'logout' })
}));
jest.mock('components/AppContent/VideoNavbar/NavbarItem', () => {
    const NavbarItem = () => <div />;
    return NavbarItem;
});
jest.mock('components/AppContent/VideoNavbar/NavbarDropdown', () => {
    const NavbarDropdown = () => <div />;
    return NavbarDropdown;
});
jest.mock('use-react-router', () => jest.fn());

const push = jest.fn();
useReactRouter.mockImplementation(() => ({
    history: {
        push
    }
}));

const mockStore = configureMockStore([thunk]);

const defaultProps = {
    disabled: false
};

const defaultStore = {
    auth: {
        userDetails: {
            firstName: 'firstName',
            lastName: 'lastName',
            roles: [
                { roleId: 1, name: 'ROLE_ADMIN' },
                { roleId: 2, name: 'ROLE_EDIT' },
                { roleId: 3, name: 'ROLE_SCAN' }
            ]
        }
    }
};

const doMount = (props = defaultProps, store = defaultStore) => {
    const reduxStore = mockStore(store);
    const component = mount(
        <Provider store={ reduxStore }>
            <MemoryRouter initialEntries={ ['/'] }>
                <VideoNavbar { ...props } />
            </MemoryRouter>
        </Provider>
    );
    return [component, store];
};

const testRendering = (component, {
    disabled = false
} = {}) => {
    expect(component.find('VideoNavbar')).toHaveLength(1);
    expect(component.find('VideoNavbar').props()).toEqual(expect.objectContaining({
        disabled
    }));
    expect(component.find('NavbarBrand')).toHaveLength(1);

    const testNavbarItem = (index, props) => {
        expect(component.find('NavbarItem').at(index).props()).toEqual(props);
    };

    if (disabled) {
        expect(component.find('NavbarItem')).toHaveLength(0);
        expect(component.find('NavbarToggler')).toHaveLength(0);
    } else {
        expect(component.find('NavbarItem')).toHaveLength(4);
        expect(component.find('NavbarToggler')).toHaveLength(1);
        expect(component.find('NavbarDropdown')).toHaveLength(1);

        testNavbarItem(0, {
            id: 'videoListLink',
            to: '/videos',
            exact: true,
            isLink: true,
            text: 'Videos'
        });
        testNavbarItem(1, {
            id: 'manageFiltersLink',
            to: '/filters',
            exact: true,
            isLink: true,
            text: 'Filters'
        });
        testNavbarItem(2, {
            id: 'userManagementLink',
            to: '/users',
            isLink: true,
            text: 'Users'
        });
        testNavbarItem(3, {
            id: 'scanDirectoryLink',
            onClick: expect.any(Function),
            text: 'Scan'
        });
    }

};

describe('VideoNavbar', () => {
    describe('rendering', () => {
        it('renders with all roles', () => {
            const [component] = doMount();
            testRendering(component);
        });

        it('renders with disabled', () => {
            const [component] = doMount({
                ...defaultProps,
                disabled: true
            });
            testRendering(component, {
                disabled: true
            });
        });

        it('renders without scanning role', () => {
            throw new Error('Finish this');
        });

        it('renders without edit role', () => {
            throw new Error('Finish this');
        });

        it('renders without admin role', () => {
            throw new Error('Finish this');
        });
    });

    describe('click actions', () => {
        beforeEach(() => {
            push.mockClear();
        });

        it('toggles the collapse open and closed', () => {
            const [component] = doMount();
            const toggle = component.find('VideoNavbar NavbarToggler');
            expect(component.find('Collapse').props()).toEqual(expect.objectContaining({
                isOpen: false
            }));
            toggle.simulate('click');
            expect(component.find('Collapse').props()).toEqual(expect.objectContaining({
                isOpen: true
            }));
        });

        it('clicks on scan directory link', async () => {
            const [component, store] = doMount();
            await component.find('NavbarItem#scanDirectoryLink').props().onClick();
            expect(store.getActions()).toEqual([
                { type: 'startFileScan' }
            ]);
            expect(push).toHaveBeenCalledWith('/scanning');
        });

        it('clicks on logout link', () => {
            const [component, store] = doMount();
            component.find('NavbarItem#logoutLink').props().onClick();
            expect(store.getActions()).toEqual([
                { type: 'logout' }
            ]);
        });
    });
});
