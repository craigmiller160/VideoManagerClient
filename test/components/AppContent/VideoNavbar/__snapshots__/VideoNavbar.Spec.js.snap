// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoNavbar renders successfully 1`] = `
<VideoNavbar
  isScanning={false}
>
  <Navbar
    color="dark"
    dark={true}
    expand="md"
    tag="nav"
  >
    <nav
      className="navbar navbar-expand-md navbar-dark bg-dark"
    >
      <Container
        tag="div"
      >
        <div
          className="container"
        >
          <NavbarBrand
            disabled={false}
            tag={[Function]}
            to="/"
          >
            <Link
              className="navbar-brand"
              disabled={false}
              to="/"
            >
              Link
            </Link>
          </NavbarBrand>
          <NavbarToggler
            onClick={[Function]}
            tag="button"
            type="button"
          >
            <button
              className="navbar-toggler"
              onClick={[Function]}
              type="button"
            >
              <span
                className="navbar-toggler-icon"
              />
            </button>
          </NavbarToggler>
          <Collapse
            appear={false}
            enter={true}
            exit={true}
            in={false}
            isOpen={false}
            mountOnEnter={false}
            navbar={true}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            tag="div"
            timeout={350}
            unmountOnExit={false}
          >
            <Transition
              appear={false}
              enter={true}
              exit={true}
              in={false}
              mountOnEnter={false}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              timeout={350}
              unmountOnExit={false}
            >
              <div
                className="collapse navbar-collapse"
                style={Object {}}
              >
                <Nav
                  className="ml-auto"
                  navbar={true}
                  tag="ul"
                  vertical={false}
                >
                  <ul
                    className="ml-auto navbar-nav"
                  >
                    <NavItem
                      tag="li"
                    >
                      <li
                        className="nav-item"
                      >
                        <NavLink
                          id="manageFilters"
                          onClick={[Function]}
                          tag="a"
                        >
                          <a
                            className="nav-link"
                            id="manageFilters"
                            onClick={[Function]}
                          >
                            Manage Filters
                          </a>
                        </NavLink>
                      </li>
                    </NavItem>
                    <NavItem
                      tag="li"
                    >
                      <li
                        className="nav-item"
                      >
                        <NavLink
                          id="scanDirectory"
                          onClick={[Function]}
                          tag="a"
                        >
                          <a
                            className="nav-link"
                            id="scanDirectory"
                            onClick={[Function]}
                          >
                            Scan Directory
                          </a>
                        </NavLink>
                      </li>
                    </NavItem>
                  </ul>
                </Nav>
              </div>
            </Transition>
          </Collapse>
        </div>
      </Container>
    </nav>
  </Navbar>
</VideoNavbar>
`;
